File: trunk/Tutorial/java/polynome_lifeCycle3/src/PolynomeUserServantProvider.java
/* =========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI.
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003.
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * =========================================================================== */


package cdmw.tutorial;


class PolynomeUserServantProvider 
    extends cdmw.tutorial.PolynomeServantProvider 
{

    private static final String POLYNOME_REP_ID = com.thalesgroup.PolynomeModule.PolynomeHelper.id();

    
    private org.omg.CORBA.ORB orb;
    private org.omg.PortableServer.POA poa;
    private cdmw.lifecycle.ObjectDeactivator objectDeactivator;

    private double counter;
    private double a;
    private double b;
    private double c;
    
    
    
    public PolynomeUserServantProvider (org.omg.CORBA.ORB orb,
                                        org.omg.PortableServer.POA poa,
                                        cdmw.lifecycle.ObjectDeactivator deactivator)
    {
        this.orb = orb;
        this.poa = poa;
        this.objectDeactivator = deactivator;
        this.counter = 0;
        this.a = 0;
        this.b = 0;
        this.c = 0;
    }


    /**
    * This method acts as a hand shaking between the Servant Provider
    * and the generated Factory implementation. It indicates whether
    * a Servant provided by this Servant Provider may incarnate multiple
    * objects (with ObjectIds provided by this Servant Provider)
    *
    * @return <B>true</B> if a servant may incarnate multiple objects,
    * <B>false</B> otherwise.
    */
    public boolean mayIncarnateMultipleObjects() {
        return false;
    }



    /**
    * Purpose:
    * <p> This method creates the oid of the HelloInterface object
    *
    *@param  object_oid string oid of object
    *
    *@return object id
    */
    public byte[] create_Polynome()
        throws org.omg.CORBA.SystemException
    {
        
        // 
        // get the equation coefs
        //
        try {
            
            java.io.BufferedReader in = 
                new java.io.BufferedReader(new java.io.InputStreamReader(System.in));
            String answer = null;
            boolean answerOK;
            
            System.out.println("Enter the coefs of polynome ax*x + bx + c = 0");
            
            do {
                System.out.print("Coef a = ");
                System.out.flush();
                answer = in.readLine();
                try {
                    a = Double.parseDouble(answer);
                    if (a == 0) { 
                        System.out.println(">>>>> Coef a cannot be null");
                    }
                    System.out.println();
                } catch (NumberFormatException e) {
                    System.out.println("This is not a number !\n");
                    a = 0;
                }
            } while (a == 0);
            
            answerOK = false;
            do {
                System.out.print("Coef b = ");
                System.out.flush();
                answer = in.readLine();
                try {
                    b = Double.parseDouble(answer);
                    answerOK = true;
                    System.out.println();
                } catch (NumberFormatException e) {
                    System.out.println("This is not a number !\n");
                }
            } while (!answerOK);
            
            answerOK = false;
            do {
                System.out.print("Coef c = ");
                System.out.flush();
                answer = in.readLine();
                try {
                    c = Double.parseDouble(answer);
                    answerOK = true;
                    System.out.println();
                } catch (NumberFormatException e) {
                    System.out.println("This is not a number !\n");
                }
            } while (!answerOK);
         
        } catch (java.io.IOException e) {
            e.printStackTrace();
            cdmw.common.Assert.check(false);
        }
        
        //
        // create an unique oid
        //
        double oidCounter;
        synchronized (this) {
            if (counter == Double.MAX_VALUE) {
                counter = 0;
            } else {
                counter++;
            }
            
            oidCounter = counter;
        }
        
        cdmw.ossupport.Timeval timeval  = cdmw.ossupport.OS.getTime();
        String polynome_oid = "POLY_" + oidCounter + "_" + timeval.getSeconds();
        
        byte[] oid = polynome_oid.getBytes();
        
        System.out.println("<<PolynomeUserServantProvider>> Polynome '" +
                           polynome_oid + "' OID created");
        
        return oid;
    }


    /**
    * Purpose:
    * <p> This method creates the HelloInterface servant associated to
    *     the oid
    *
    *@return servant object
    */    
    public org.omg.PortableServer.Servant lookupServant(byte[] oid) 
        throws org.omg.PortableServer.ForwardRequest 
    {
        com.thalesgroup.PolynomeModule.PolynomeOperations delegate;
        
        //
        // Create implementation object 
        //
        
        // compute the discriminant Delta = b*b - 4*a*c
        double delta = b*b - 4*a*c;
        
        // if delta is negative : no roots
        if (delta < 0)
        {
            NoRootsBinomeImpl binome = new NoRootsBinomeImpl (objectDeactivator);
                
            delegate = binome;
        
            System.out.println("<<PolynomeUserServantProvider>> NoRootsBinomeImpl servant created");
        }
        
        // else if delta is zero : only one root
        else if (delta == 0)
        {
            SingleRootBinomeImpl binome = 
                 new SingleRootBinomeImpl (objectDeactivator, a, b);
                
            delegate = binome;
        
            System.out.println("<<PolynomeUserServantProvider>> SingleRootBinomeImpl servant created");
        }
        
        // else delta is positive : two roots
        else
        {
            TwoRootsBinomeImpl binome = 
                  new TwoRootsBinomeImpl (objectDeactivator, a, b, delta);
                          
            delegate = binome;
        
            System.out.println("<<PolynomeUserServantProvider>> TwoRootsBinome_impl servant created");
        }        

        // NOTE: The servant is a TIE (see note into Impls)
        org.omg.PortableServer.Servant servant = 
            new com.thalesgroup.PolynomeModule.PolynomePOATie(delegate, poa);
        
        // return the servant reference of object
        return servant;
    }


    public void releaseServant(byte[] oid,
                               org.omg.PortableServer.POA poa,
                               org.omg.PortableServer.Servant servant) 
    {
        System.out.println("<<PolynomeUserServantProvider>> Polynome servant reference removed.");
    }


    public String getSupportedInterface(byte[] oid)
    {
        return POLYNOME_REP_ID;    
    }


        
}

File: Tutorial/c++/polynome_lifeCycle3/src/PolynomeUserServantProvider.cpp
/* =========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI.
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003.
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * =========================================================================== */


#include <string>
#include <cctype>
#include <algorithm>
#include <sstream>


#include <Foundation/ossupport/OS.hpp>
#include <Foundation/osthreads/MutexGuard.hpp>
#include <LifeCycle/lifecycle/LifeCycleObjectBase.hpp>


#include "PolynomeUserServantProvider.hpp"
#include "NoRootsBinome_impl.hpp"
#include "SingleRootBinome_impl.hpp"
#include "TwoRootsBinome_impl.hpp"


namespace 
{
    const char* POLYNOME_REP_ID = "IDL:thalesgroup.com/PolynomeModule/Polynome:1.0";
};



namespace Cdmw
{
namespace Tutorial
{



PolynomeUserServantProvider::~PolynomeUserServantProvider()
    throw()
{
}



PolynomeUserServantProvider::PolynomeUserServantProvider (
                   CORBA::ORB_ptr orb,
                   PortableServer::POA_ptr poa,
                   const Cdmw::LifeCycle::ObjectDeactivator & deactivator,
                   int & argc, char** argv)
  throw(CORBA::SystemException)
  
  : m_orb(CORBA::ORB::_duplicate(orb)),
    m_poa(PortableServer::POA::_duplicate(poa)),
    m_object_deactivator(deactivator)
{
    m_counter=0L;
}



PortableServer::ObjectId* PolynomeUserServantProvider::create_Polynome ()
    throw ()
{
    // 
    // get the equation coefs
    //
    std::cout << "Enter the coefs of polynome ax*x + bx + c = 0" 
              << std::endl;
    
    m_a=0;
    while (m_a == 0)
    {
        std::cout << "Coef a = ";
        std::cin >> m_a;
        std::cout << std::endl;
                    
        if (m_a == 0)
        {
            std::cout << ">>>>> Coef a cannot be null" << std::endl;
            std::cout << std::endl;
        }
    }
    
    std::cout << "Coef b = ";
    std::cin >> m_b;
    std::cout << std::endl;
    
    std::cout << "Coef c = ";
    std::cin >> m_c;
    std::cout << std::endl;
    
    //
    // create an unique oid
    //
    
    unsigned long oid_counter;
    
    {
        const unsigned long MAX_COUNTER = 0xFFFFFFFF;
        
        CDMW_MUTEX_GUARD(m_count_mutex);
        
        if (m_counter == MAX_COUNTER)
        {
            m_counter = 0L;
        }
        else
        {
            m_counter++;
        }
        
        oid_counter = m_counter;
    }
    
    OsSupport::OS::Timeval timeval = OsSupport::OS::get_time();
    
    std::stringstream ostr;

    ostr << "POLY_" << oid_counter << "_" << timeval.seconds;
    
	PortableServer::ObjectId_var oid = 
	     PortableServer::string_to_ObjectId(ostr.str().c_str());
	                                       
	
   	std::cout << "<<PolynomeUserServantProvider>> Polynome '" 
   	          << ostr.str().c_str() <<"' OID created" << std::endl;
	
	return oid._retn();
}



PortableServer::Servant 
PolynomeUserServantProvider::lookup_servant(const PortableServer::ObjectId& oid) 
    throw(PortableServer::ForwardRequest,
          CORBA::SystemException)
{
    PortableServer::Servant servant;

    //
    // Create implementation object 
    //
    
    // compute the discriminant Delta = b*b - 4*a*c
    double delta = m_b*m_b - 4*m_a*m_c;
    
    // if delta is negative : no roots
    if (delta < 0)
    {
        NoRootsBinome_impl* p_binome = new NoRootsBinome_impl (m_object_deactivator);
            
        servant = p_binome;
    
        std::cout << "<<PolynomeUserServantProvider>> NoRootsBinome_impl servant created" << std::endl;
    }
    
    // else if delta is zero : only one root
    else if (delta == 0)
    {
        SingleRootBinome_impl* p_binome = 
             new SingleRootBinome_impl (m_object_deactivator,m_a,m_b);
            
        servant = p_binome;
    
        std::cout << "<<PolynomeUserServantProvider>> SingleRootBinome_impl servant created" << std::endl;
    }
    
    // else delta is positive : two roots
    else
    {
        TwoRootsBinome_impl* p_binome = 
              new TwoRootsBinome_impl (m_object_deactivator,m_a,m_b,delta);
                      
        servant = p_binome;
    
        std::cout << "<<PolynomeUserServantProvider>> TwoRootsBinome_impl servant created" << std::endl;
    }
    
    // return the servant reference of object
    return servant;
}


void PolynomeUserServantProvider::release_servant(
                     const PortableServer::ObjectId& oid,
                     PortableServer::POA_ptr poa,
                     PortableServer::Servant servant) 
    throw(CORBA::SystemException)
{            
    servant->_remove_ref();
        
    std::cout << "<<PolynomeUserServantProvider>> Polynome servant reference removed." << std::endl;
}



std::string PolynomeUserServantProvider::get_supported_interface(
                         const PortableServer::ObjectId& oid) const
    throw(CORBA::OBJECT_NOT_EXIST)
{
	std::string result = POLYNOME_REP_ID;
            
    return result;    
}



        
}; // End namespace Tutorial
}; // End namespace Cdmw
